Sources/_StringProcessing/Regex/Match.swift

```diff
@@ -37,33 +41,48 @@ extension Regex.Match {
       )
       
       let output = AnyRegexOutput(
-        input: anyRegexOutput.input,
+        input: input,
         elements: [wholeMatchCapture] + anyRegexOutput._elements
       )
       
       return output as! Output
     } else if Output.self == Substring.self {
       // FIXME: Plumb whole match (`.0`) through the matching engine.
-      return anyRegexOutput.input[range] as! Output
-    } else if anyRegexOutput.isEmpty, value != nil {
+      return input[range] as! Output
+    } else if anyRegexOutput.isEmpty, let value {
       // FIXME: This is a workaround for whole-match values not
       // being modeled as part of captures. We might want to
       // switch to a model where results are alongside captures
-      return value! as! Output
+      return value as! Output
     } else {
       guard value == nil else {
         fatalError("FIXME: what would this mean?")
       }
       let typeErasedMatch = anyRegexOutput.existentialOutput(
-        from: anyRegexOutput.input[range]
+        from: input[range]
       )
       return typeErasedMatch as! Output
     }
   }
 
+  var wholeMatchType: Any.Type {
+    value.map { type(of: $0) } ?? Substring.self
+  }
+
   /// Accesses a capture by its name or number.
   public subscript<T>(dynamicMember keyPath: KeyPath<Output, T>) -> T {
-    output[keyPath: keyPath]
+    // Note: We should be able to get the element offset from the key path
+    // itself even at compile time. We need a better way of doing this.
+    guard let outputTupleOffset = MemoryLayout.tupleElementIndex(
```

Comment:
When we store the whole match inside the capture list, we will be able to precompute these offsets for all elements in a `Regex.Match`. I'll do that next.